File: libraries/dotnet/UrcConverter/Sources/Osu/Convert.fs
namespace UrcConverter.Sources.Osu

module Convert =

    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources
    open UrcConverter.Sources.Osu.Model

    let private keyMin, keyMax = 1, 18
    let private judgmentRates = [ 100.0; 100.0; 66.67; 33.33; 16.67; 0.0 ]

    let convert (beatmap: Beatmap) : Result<Chart, UrcError> =
        result {
            if beatmap.Mode <> 3 then
                return!
                    Result.Error(
                        UrcError.UnsupportedVersion(1, $"unsupported game mode: {beatmap.Mode}")
                    )

            match beatmap.CircleSize with
            | None -> return! Result.Error(UrcError.Syntax(1, "missing CircleSize"))
            | Some circleSize when circleSize <> truncate circleSize ->
                return!
                    Result.Error(UrcError.Syntax(1, $"CircleSize must be an integer: {circleSize}"))
            | Some circleSize ->
                let keys = int circleSize

                if keys < keyMin || keys > keyMax then
                    return! Result.Error(UrcError.Syntax(1, $"CircleSize out of range: {keys}"))

                if beatmap.TimingPoints |> List.exists (fun point -> point.BeatLength = 0.0) then
                    return! Result.Error(UrcError.Syntax(1, "timing point with zero beat length"))

                let firstNoteTime =
                    match beatmap.HitObjects with
                    | [] -> 0
                    | objects -> objects |> List.map _.Time |> List.min

                let bpmPoints =
                    beatmap.TimingPoints
                    |> List.choose (fun point ->
                        if point.Uninherited then
                            Some(point.Time, 60000.0 / point.BeatLength, point.Meter)
                        else
                            None)

                let! timing =
                    Shared.buildTiming
                        bpmPoints
                        (beatmap.TimingPoints
                         |> List.choose (fun point ->
                             if not point.Uninherited then
                                 Some(point.Time, -100.0 / point.BeatLength)
                             else
                                 None))
                        firstNoteTime
                        ".osu"
                        (Shared.firstDownbeatAfter bpmPoints firstNoteTime)

                let! notes =
                    beatmap.HitObjects
                    |> List.traverseResultM (fun obj ->
                        let lane = min (max (obj.X * keys / 512) 0) (keys - 1)

                        match obj.IsHold, obj.EndTime with
                        | true, Some endTime when endTime < obj.Time ->
                            Result.Error(
                                UrcError.Syntax(
                                    1,
                                    $"hold ends before it starts: {endTime} < {obj.Time}"
                                )
                            )
                        | true, Some endTime ->
                            Result.Ok
                                [
                                    {
                                        TimestampMs = obj.Time - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.LS
                                    }
                                    {
                                        TimestampMs = endTime - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.LE
                                    }
                                ]
                        | true, None -> invalidOp "hold hit object must have an end time"
                        | false, _ ->
                            Result.Ok
                                [
                                    {
                                        TimestampMs = obj.Time - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.N
                                    }
                                ])
                    |> Result.map List.concat

                do! Shared.checkHoldOverlap notes

                let judgment =
                    beatmap.OverallDifficulty
                    |> Option.map (fun od ->
                        let windows =
                            16.5
                            :: [
                                for baseline in [ 64.0; 97.0; 127.0; 151.0; 188.0 ] ->
                                    baseline - 3.0 * od + 0.5
                            ]

                        {
                            Windows = windows
                            Rates = judgmentRates
                        })

                let title = beatmap.TitleUnicode |> Option.orElse beatmap.Title
                let artist = beatmap.ArtistUnicode |> Option.orElse beatmap.Artist

                let metadata =
                    [
                        ("Title", title)
                        ("Artist", artist)
                        ("Creator", beatmap.Creator)
                        ("Version", beatmap.Version)
                    ]

                let missing =
                    metadata
                    |> List.choose (fun (name, value) ->
                        match value with
                        | Some value when value <> "" -> None
                        | _ -> Some name)

                let missingText = missing |> String.concat ", "

                if missing <> [] then
                    return! Result.Error(UrcError.Syntax(1, $"missing metadata: {missingText}"))

                match title, artist, beatmap.Creator, beatmap.Version with
                | Some title, Some artist, Some creator, Some version ->
                    return
                        {
                            FormatVersion = { Major = 1; Minor = 1 }
                            Metadata =
                                {
                                    Original = "osu!mania"
                                    Title = title
                                    Artist = artist
                                    Creator = creator
                                    Version = version
                                }
                            Judgment = judgment
                            Layout =
                                {
                                    Keys = keys
                                    SpecialKeys = 0
                                    SpecialLanes = None
                                }
                            TimingPoints = timing
                            Notes = notes
                        }
                | _ ->
                    return!
                        Result.Error(UrcError.Syntax(1, $"missing metadata: {missingText}"))
        }

File: libraries/python/src/urc_converter/sources/osu/convert.py
"""Mapper from the osu!mania source model onto a URC chart."""

import math

from ...error import UrcError
from ...model import Chart, Judgment, Layout, Metadata, Note, NoteType, Version
from .._shared import build_timing, check_hold_overlap, first_downbeat_after
from .model import OsuBeatmap

_JUDGMENT_RATES = [100.0, 100.0, 66.67, 33.33, 16.67, 0.0]
_KEY_MIN, _KEY_MAX = 1, 18


def convert_osu(beatmap: OsuBeatmap) -> Chart:
	"""Map an osu!mania beatmap onto a URC chart."""
	if beatmap.mode != 3:
		raise UrcError("unsupported-version", 1, f"unsupported game mode: {beatmap.mode}")

	if beatmap.circle_size is None:
		raise UrcError("syntax", 1, "missing CircleSize")

	if beatmap.circle_size != int(beatmap.circle_size):
		raise UrcError("syntax", 1, f"CircleSize must be an integer: {beatmap.circle_size}")

	keys = int(beatmap.circle_size)
	if not _KEY_MIN <= keys <= _KEY_MAX:
		raise UrcError("syntax", 1, f"CircleSize out of range: {keys}")

	if any(point.beat_length == 0 for point in beatmap.timing_points):
		raise UrcError("syntax", 1, "timing point with zero beat length")

	first_note_time = min((obj.time for obj in beatmap.hit_objects), default=0)

	bpm_points = [
		(point.time, 60000.0 / point.beat_length, point.meter, 4)
		for point in beatmap.timing_points
		if point.uninherited
	]
	timing = build_timing(
		bpm_points=bpm_points,
		sv_points=[
			(point.time, -100.0 / point.beat_length)
			for point in beatmap.timing_points
			if not point.uninherited
		],
		first_note_time=first_note_time,
		source=".osu",
		measure_anchor_ms=first_downbeat_after(bpm_points, first_note_time),
	)

	notes: list[Note] = []
	for obj in beatmap.hit_objects:
		lane = min(max(math.floor(obj.x * keys / 512), 0), keys - 1)

		if obj.is_hold:
			if obj.end_time < obj.time:
				raise UrcError(
					"syntax", 1, f"hold ends before it starts: {obj.end_time} < {obj.time}"
				)
			notes.append(Note(timestamp_ms=obj.time - first_note_time, lane=lane, type=NoteType.LS))
			notes.append(
				Note(timestamp_ms=obj.end_time - first_note_time, lane=lane, type=NoteType.LE)
			)
		else:
			notes.append(Note(timestamp_ms=obj.time - first_note_time, lane=lane, type=NoteType.N))

	check_hold_overlap(notes)

	judgment = None
	if beatmap.overall_difficulty is not None:
		od = beatmap.overall_difficulty
		windows = [16.5] + [base - 3 * od + 0.5 for base in (64, 97, 127, 151, 188)]
		judgment = Judgment(windows=windows, rates=list(_JUDGMENT_RATES))

	title = beatmap.title_unicode or beatmap.title
	artist = beatmap.artist_unicode or beatmap.artist
	missing = [
		label
		for label, value in (
			("Title", title),
			("Artist", artist),
			("Creator", beatmap.creator),
			("Version", beatmap.version),
		)
		if not value
	]
	if missing:
		raise UrcError("syntax", 1, f"missing metadata: {', '.join(missing)}")

	return Chart(
		format_version=Version(major=1, minor=1),
		metadata=Metadata(
			original="osu!mania",
			title=title,
			artist=artist,
			creator=beatmap.creator,
			version=beatmap.version,
		),
		judgment=judgment,
		layout=Layout(keys=keys, special_keys=0, special_lanes=None),
		timing=timing,
		notes=notes,
	)
